test/testutils/curl_request.go

```
@@ -0,0 +1,215 @@
+	c.service = service
+	return c
+}
+
+func (c *CurlRequestBuilder) WithSni(sni string) *CurlRequestBuilder {
+	c.sni = sni
+	return c
+}
+
+func (c *CurlRequestBuilder) WithCaFile(caFile string) *CurlRequestBuilder {
+	c.caFile = caFile
+	return c
+}
+
+func (c *CurlRequestBuilder) WithPath(path string) *CurlRequestBuilder {
+	c.path = path
+	return c
+}
+
+func (c *CurlRequestBuilder) WithPostBody(body string) *CurlRequestBuilder {
+	return c.WithBody(body).WithContentType("application/json")
+}
+
+func (c *CurlRequestBuilder) WithBody(body string) *CurlRequestBuilder {
+	c.body = body
+	return c
+}
+
+func (c *CurlRequestBuilder) WithContentType(contentType string) *CurlRequestBuilder {
+	return c.WithHeader("Content-Type", contentType)
+}
+
+func (c *CurlRequestBuilder) WithHost(host string) *CurlRequestBuilder {
+	return c.WithHeader("Host", host)
+}
+
+func (c *CurlRequestBuilder) WithHeader(key, value string) *CurlRequestBuilder {
+	c.headers[key] = value
+	return c
+}
+
+func (c *CurlRequestBuilder) WithScheme(scheme string) *CurlRequestBuilder {
+	c.scheme = scheme
+	return c
+}
+
+// WithArgs allows developers to append arbitrary args to the CurlRequestBuilder
+// This should mainly be used for debugging purposes. If there is an argument that the builder
+// doesn't yet support, it should be added explicitly, to make it easier for developers to utilize
+func (c *CurlRequestBuilder) WithArgs(args []string) *CurlRequestBuilder {
+	c.additionalArgs = args
+	return c
+}
+
+func (c *CurlRequestBuilder) errorIfInvalid() error {
+	if c.service == "" {
+		return errors.New("service is empty, but required")
+	}
+
+	return nil
+}
+
+func (c *CurlRequestBuilder) BuildArgs() []string {
+	ginkgo.GinkgoHelper()
+
+	if err := c.errorIfInvalid(); err != nil {
+		// We error loudly here
+		// These types of errors are intended to prevent developers from creating resources
+		// which are semantically correct, but lead to test flakes/confusion
+		ginkgo.Fail(err.Error())
```

It looks like we're only using ginkgo here to fail- can we just return the error instead?

---

We could, and I'm happy to do that. My hope was basically that we know there are cases that should never pass, and so instead of deferring the assertion to the caller of this function, we could bake the logic directly in. 

What do you think about the tradeoff of baking the logic into this component versus delegating it to the user?

---

Yeah I think it depends- it looks like the `errorIfInvalid()` is called only when the `c.service` is unset? But then the`DefaultCurlRequestBuilder` doesn't have the service set? It looks like setting the service field is already user controlled, not something that's set by the builder, so it's already the user's responsibility. 

If `c.service` was set by the builder based on some env/dependency, I think this pattern of erroring loudly makes sense. It does have the tradeoff that we're trying our test util code to ginkgo.